SAML Integration Tests

anijain-Amazon opened the pull request:
### Description
Adds integration tests using selenium web driver for SAML Authentication flows

### Category
Integration Tests 

### What is the old behavior before changes and new behavior after changes?
Integration test were not included. 

### Issues Resolved
Integration tests for fixes https://github.com/opensearch-project/security-dashboards-plugin/pull/1058 and https://github.com/opensearch-project/security-dashboards-plugin/pull/1039

Duplicate of the previous PR https://github.com/opensearch-project/security-dashboards-plugin/pull/1044

### Testing
Integration tests are present in the file `saml_auth.test.ts`

### Check List
- [x] New functionality includes testing
- [x] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

anijain-Amazon commented:
@peternied The integration tests keep failing because of an issue we have observed during testing as well. 
[FailedWorkflow](https://github.com/opensearch-project/security-dashboards-plugin/runs/8101889398?check_suite_focus=true)

For local testing we performed the following steps : 
- Start and stop the OpenSearch Dashboard Server using `yarn start`
- Run the integration tests using `yarn test:jest_server`

If we don't run `yarn start` after a fresh checkout, then the Selenium Web-driver throws up an error ( attached SS ) after logging in.

<img width="1484" alt="Screenshot 2022-08-30 at 2 43 41 PM" src="https://user-images.githubusercontent.com/110471048/187554502-3c36f274-a77b-47e3-ac56-d743590f4bc5.png">

A naive solution is to start and stop the OpenSearch Dashboard Server [before this](https://github.com/opensearch-project/security-dashboards-plugin/blob/main/.github/workflows/integration-test.yml#L112)

Could you share some insights on what exactly could be causing this ?

anijain-Amazon commented:
I have identified the issue for the integration tests failing in Github Actions workflow. 

Enabled verbose logging for OSD Server.

```
logging: {
  silent: false,
  verbose: true
}
```

Checked out the security-dashboard-plugin and ran `yarn osd bootstrap`

Ran the jest Integration tests using the command `yarn test:jest_server` with headless mode disabled for the Selenium web-driver. 

When the tests ran, all the ITs failed as after SAML login the error page was thrown. Upon checking the logs, I could see lot of calls to fetch plugin bundles were failing with the `404` response code. 

<img width="1391" alt="Screenshot 2022-08-31 at 11 12 44 PM" src="https://user-images.githubusercontent.com/110471048/187744455-83f93898-25f2-416a-99b4-db724d2f00f3.png">

Checked around OSD Server code base and found that [build_opensearch_dashboards_platform_plugins](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/scripts/build_opensearch_dashboards_platform_plugins.js) gets invoked for generating these plugins when `yarn start` command is called. 

So, ran the command `node scripts/build_opensearch_dashboards_platform_plugins.js` after `yarn osd bootstrap` before starting the Integration tests. The entire suite of selenium web-driver based tests ran without any hiccups. 

In order, to make the same run in Github Actions, I think we can add the following line before [this](https://github.com/opensearch-project/security-dashboards-plugin/blob/2.1/.github/workflows/integration-test.yml#L111)
```
node ./OpenSearch-Dashboards/scripts/build_opensearch_dashboards_platform_plugins.js
```
This will ensure that all integration tests using selenium web driver with security-dashboards-plugin work seamlessly in the future.

cliu123 commented on .github/workflows/integration-test.yml:
```diff
@@ -103,10 +103,20 @@ jobs:
+          echo -e "\n\nPrinting the Settings"
+          curl -XGET https://localhost:9200/_cluster/settings?include_defaults=true -u 'admin:admin' -k
+          echo -e "\nFetching Docker Logs for OS Server\n"
+          docker logs oss
```
Are `\n`s needed from line 117 - 122?

anijain-Amazon replied:
These are added for debugging, will remove once the ITs succeed.

cliu123 commented on server/backend/opensearch_security_client.ts:
```diff
@@ -157,6 +157,9 @@ export class SecurityClient {
       //     requestId="<request_id>"
       //   '
+
+      console.log('Recieved error from authinfo API', error);
```
Please remove if this is for debugging.

anijain-Amazon replied:
These are added for debugging, will remove once the ITs succeed.

cliu123 commented on test/helper/cookie.ts:
```diff
@@ -20,7 +20,7 @@ import { AUTHORIZATION_HEADER_NAME } from '../constant';
 export function extractAuthCookie(response: Response) {
   const setCookieHeaders = response.header['set-cookie'] as string[];
-  let securityAuthCookie: string;
+  let securityAuthCookie: string | null = null;
```
Does `securityAuthCookie` need to be `null` here? The [return value](https://github.com/opensearch-project/security-dashboards-plugin/blob/07688fab15a6ddc2ccde86597144b67eab59d265/test/helper/cookie.ts#L30) can be `undefined`.

devardee replied:
yes, since null can valid value for securityAuthCookie

cliu123 commented on test/jest_integration/saml_auth.test.ts:
```diff
@@ -0,0 +1,340 @@
+      }
+    );
+
+    console.log('Starting OpenSearchDashboards server..');
```
Was this log added on purpose or only for debugging?

cliu123 commented on test/jest_integration/saml_auth.test.ts:
```diff
@@ -0,0 +1,340 @@
+    await driver.wait(until.elementsLocated(By.xpath(pageTitleXPath)), 10000);
+
+    const cookie = await driver.manage().getCookies();
+    expect(cookie.length).toEqual(2);
```
Is the length of the cookie a constant here? alway equals 2?

devardee replied:
yes, one cookie for set by the idp and one cookie set by the security dashboard plugin. So in total there should be two cookies

cwperks replied:
It looks like the confusion here is in the variable name `cookie`. The call to `getCookies` is returning all cookies that the browser encounters which is multiple cookies in this case. Perhaps we should rename this variable to `cookies`?

peternied reviewed (commented):
>In order, to make the same run in Github Actions, I think we can add the following line before [this](https://github.com/opensearch-project/security-dashboards-plugin/blob/2.1/.github/workflows/integration-test.yml#L111)
>`node ./OpenSearch-Dashboards/scripts/build_opensearch_dashboards_platform_plugins.js`
>This will ensure that all integration tests using selenium web driver with security-dashboards-plugin work seamlessly in the future.

Sure, if we need to update how the configuration is performed we should do so, it looks this added this to the current changes 👍 

In the CI run it looks like there are some failures - it seems like there was several API call failures ` Recieved error from authinfo API StatusCodeError {...` that might indicate there was a problem during the test setup

anijain-Amazon commented on test/jest_integration/saml_auth.test.ts:
```diff
@@ -0,0 +1,340 @@
+    } catch (error) {
+      console.log('Got an error!!', error.stack);
+    }
+    console.log('The Config Response is : ' + JSON.stringify(config));
```
We will remove it post integration tests run successfully.

anijain-Amazon commented:
@peternied @cliu123 Addressed most of the comments and the Integration tests are running as expected. 

There is an [expected failure](https://github.com/devardee/security-dashboards-plugin/runs/8193919806?check_suite_focus=true#step:12:1464) for the test case of Tenancy getting retained after logout in SAML because the [fix](https://github.com/opensearch-project/security-dashboards-plugin/pull/1090) is not present in this PR.

DarshitChanpura commented on .github/workflows/integration-test.yml:
```diff
@@ -11,38 +11,9 @@ jobs:
-          ENTRYPOINT /docker-host/os-ep.sh
-          EOF
-          docker run -d -p 9200:9200 -p 9600:9600 -i opensearch-test:latest
+          docker run -d --network=host -e "discovery.type=single-node" -e "plugins.security.unsupported.restapi.allow_securityconfig_modification=true" -i opensearchproject/opensearch:2.2.0
```
this should point to `opensearch:latest` ?

peternied commented on .github/workflows/integration-test.yml:
```diff
@@ -11,38 +11,9 @@ jobs:
     name: Run integration tests
     runs-on: ubuntu-latest
     steps:
```
We need to review the deletion of the manual setup of opensearch and dashboards.  This is the only way we can test pre-released features from the security backend or dashboards.

For more context see https://github.com/opensearch-project/security-dashboards-plugin/issues/886

peternied commented:
@anijain-Amazon This looks really close, can you disable the test that you expect to fail with a reference to the PR so the CI run can pass?

anijain-Amazon commented:
> @anijain-Amazon This looks really close, can you disable the test that you expect to fail with a reference to the PR so the CI run can pass?

Do you want me to comment it for one run ?

anijain-Amazon commented:
@peternied Fixed the merge conflicts and commented out the failing IT. All tests are passing now 
[Integration Test Run](https://github.com/devardee/security-dashboards-plugin/actions/runs/3050426213)
[Unit Test Run](https://github.com/devardee/security-dashboards-plugin/actions/runs/3050426205)
[PreRequisite Test Run](https://github.com/devardee/security-dashboards-plugin/actions/runs/3050426208)

peternied commented:
@anijain-Amazon Still seeing this failure, if this is expected until the bug fix, please disable with `.skip` [[SO](https://stackoverflow.com/a/48125240/533057)].
```
  ● start OpenSearch Dashboards server › Tenancy persisted after Logout in SAML

    expect(received).toEqual(expected) // deep equality

    Expected: "Global"
    Received: "Private"

      328 |     await driver.quit();
      329 |
    > 330 |     expect(tenantName).toEqual('Global');
          |                        ^
      331 |   });
      332 | });
      333 |

      at Object.<anonymous> (plugins/security-dashboards-plugin/test/jest_integration/saml_auth.test.ts:330:24)
          at runMicrotasks (<anonymous>)
```

cliu123 commented on package.json:
```diff
@@ -17,19 +17,23 @@
+    "@types/hapi__wreck": "^15.0.1",
+    "gulp-rename": "2.0.0",
+    "saml-idp": "^1.2.1",
+    "selenium-webdriver": "^4.0.0-alpha.7",
```
@peterzhuamazon @bbarani Are there any concerns on building selenium into release?

cliu123 replied:
@anijain-Amazon Would it be possible to add these tests in [opensearch-dashboards-functional-test](https://github.com/opensearch-project/opensearch-dashboards-functional-test/tree/main/cypress/integration/plugins/security) and implement with Cypress? There has already been a well-automated workflow in place where developers can easily implement tests and the tests run with each build. With the workflow, developers do not have to worry about setting up clusters and configurations. They can just focus on implementing tests and the workflow takes care of everything else. In comparison, Selenium would be a brand new test framework that would require some changes in Infra for release build.

devardee replied:
@cliu123 currently cypress does not support saml auth workflow https://github.com/cypress-io/cypress/issues/5397, so we went with selenium

cliu123 replied:
I see. Thanks for the information!
@kavilla @peterzhuamazon Please be aware of this new test dependency Selenium.

anijain-Amazon commented:
> @anijain-Amazon Still seeing this failure, if this is expected until the bug fix, please disable with `.skip` [[SO](https://stackoverflow.com/a/48125240/533057)].
> 
> ```
>   ● start OpenSearch Dashboards server › Tenancy persisted after Logout in SAML
> 
>     expect(received).toEqual(expected) // deep equality
> 
>     Expected: "Global"
>     Received: "Private"
> 
>       328 |     await driver.quit();
>       329 |
>     > 330 |     expect(tenantName).toEqual('Global');
>           |                        
>       331 |   });
>       332 | });
>       333 |
> 
>       at Object.<anonymous> (plugins/security-dashboards-plugin/test/jest_integration/saml_auth.test.ts:330:24)
>           at runMicrotasks (<anonymous>)
> ```

@peternied Please see the links I have attached above where I had commented the test. This failure is happening because I re-enabled the tests again. Adding the Integration test links again


https://github.com/devardee/security-dashboards-plugin/actions/runs/3050426213

anijain-Amazon commented:
@peternied I have commented out the tests again. 
Please let me know when you have verified it from your end. 
I will enable it again.

cwperks commented on test/jest_integration/saml_auth.test.ts:
```diff
@@ -0,0 +1,334 @@
+
+    await driver.wait(
+      until.elementsLocated(
+        By.xpath('/html/body/div[1]/div/div/div/div[2]/div/main/div[1]/span/button/span')
```
This is quite a long path to the element and its not clear what element this is waiting for. Is this something with the control of the security-dashboards-plugin or would we need to add an id to this span in OSD?

cwperks commented on test/jest_integration/saml_auth.test.ts:
```diff
@@ -0,0 +1,334 @@
+});
+
+function getDriver() {
+  return new Builder().forBrowser('firefox').setFirefoxOptions(new Options().headless());
```
Would it make sense at all to parameterize `firefox` to be able to run these tests with a different browser?

cwperks commented on test/jest_integration/saml_auth.test.ts:
```diff
@@ -0,0 +1,334 @@
+        },
+      });
+    } catch (error) {
+      console.log('Got an error!!', error.stack);
```
Should the test be aborted at this point since the saml config was not loaded successfully?

cwperks commented on package.json:
```diff
@@ -17,6 +17,7 @@
     "lint:es": "node ../../scripts/eslint",
     "lint:style": "node ../../scripts/stylelint",
     "lint": "yarn run lint:es && yarn run lint:style",
+    "pretest:jest_server": "node ./test/jest_integration/runIdpServer.js &",
```
I'm curious as well, can we spin this up and tear it down as needed for a test or suite of tests?

anijain-Amazon commented:
> @[anijain-Amazon](https://github.com/anijain-Amazon) anijain-Amazon dismissed [peternied](https://github.com/peternied)’s [stale review](https://github.com/opensearch-project/security-dashboards-plugin/pull/1088#pullrequestreview-1109412107) via [847bc2d](https://github.com/opensearch-project/security-dashboards-plugin/commit/847bc2d4c538031c4d68d8d3174a54454378468d) [1 hour ago](https://github.com/opensearch-project/security-dashboards-plugin/pull/1088#event-7397497219)

@peternied Was fixing the comments from @cwperks which led to your approval being reverted.

cwperks commented on test/jest_integration/saml_auth.test.ts:
```diff
@@ -0,0 +1,334 @@
+});
+
+function getDriver() {
+  return new Builder().forBrowser('firefox').setFirefoxOptions(new Options().headless());
```
Also making headless into a parameter for a developer writing tests to run this headed or headless could be useful.

cwperks reviewed (dismissed):
@devardee @anijain-Amazon Thank you for this contribution to automate SAML testing so we can better proactively identify breaking changes to the SAML integration. Good work figuring out the issue with `node scripts/build_opensearch_dashboards_platform_plugins.js`, after reading this PR I was able to run it locally. 

For someone trying to perform the tests locally there is quite a bit of setup. Would you be able to add a section in the DEVELOPER_GUIDE.md to provide guidance in setup?

Thank you for removing references to long XPaths, I think this will be more robust with using element identifiers.

anijain-Amazon commented:
> @[anijain-Amazon](https://github.com/anijain-Amazon) anijain-Amazon dismissed stale reviews from [cwperks](https://github.com/cwperks) and [peternied](https://github.com/peternied) via [9af274a](https://github.com/opensearch-project/security-dashboards-plugin/commit/9af274a5146c97c24f6529aa628e2b62f860be51) [10 hours ago](https://github.com/opensearch-project/security-dashboards-plugin/pull/1088#event-7398548871)

@peternied @cwperks Your approvals got stale after making change to DEVELOPER Guide. Please provide it again and merge the PR.

anijain-Amazon commented:
> Note, this PR can't be merged until CI is passing

I think the only to merge this PR would be to disable the test and enable it again in this [PR](https://github.com/opensearch-project/security-dashboards-plugin/pull/1058)

anijain-Amazon commented:
@peternied @cwperks I have commented out the [assertion](https://github.com/opensearch-project/security-dashboards-plugin/pull/1088/commits/f247dbbd4b6885fb05493d15d234908008c68639) for failing IT. 
The Integration tests are passing now. Will uncomment the same after rebase in the [PR](https://github.com/opensearch-project/security-dashboards-plugin/pull/1058) which has the fix. 

Please provide the approvals again and merge this PR.

cwperks commented:
Backport to 2.x failed - manually created backport PR: https://github.com/opensearch-project/security-dashboards-plugin/pull/1104

cwperks commented:
Automatic backport failed because this PR did not get the backport 2.x label: https://github.com/opensearch-project/security-dashboards-plugin/pull/1039
